Error while compiling branch "addControlModes"

I checked out this branch, I run make uninstall and I removed the build directory. Then, I tried to compile the repo. I got the following error:

 ```
Building CXX object CMakeFiles/bridge_ihmc_ors.dir/yarp_plugin_bridge_ihmc_ors.cpp.o
In file included from /home/gabri/Software/gitHub/robotology-playground/ihmc-ors-yarp/build/yarp_plugin_bridge_ihmc_ors.cpp:10:0:
/home/gabri/Software/gitHub/robotology-playground/ihmc-ors-yarp/src/BridgeIHMCORS.h:142:10: error: ‘mutex’ in namespace ‘std’ does not name a type
     std::mutex m_jointORSControlModeMutex;

```
Isn't it because `mutex` is actually a `yarp` and not `std` class?

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -109,7 +109,7 @@ option(BUILD_TESTING "Create tests using CMake" OFF)
 if(BUILD_TESTING)
   include(CTest)
   enable_testing()
-  # add_subdirectory(tests)
+  add_subdirectory(tests)
 endif()
 
 # Add the "uninstall" target

--- a/README.md
+++ b/README.md
@@ -98,14 +98,19 @@ By default, after the start the device send the RobotFeedback message as a UDP d
 on the port `feedback-port-number` (default value: `9970`). A separate threads listen for a UDP datagram containing the RobotDesireds message 
  on the `desired-address` IP address  on the port `desired-port-number` (default value: `9980`)
  
-![state](https://user-images.githubusercontent.com/1857049/32063969-e05ff15c-ba78-11e7-9f20-5cb08d86f54d.png)
+### Joint State Behaviour 
+![state](https://user-images.githubusercontent.com/1857049/32371398-f7be7274-c090-11e7-8bc7-fd539845c8cf.png)
 
-The bridge starts in the OFF state. As soon as it receives a RobotDesireds message, it switches to the ON state. While entering in the ON state, it switches all the YARP control modes of the joint handled by the bridge to `VOCAB_CM_TORQUE`.
+The bridge starts all the joints in the `NOT_ENABLED`  state, and the state of each joint evolve indipendently from the rest of the joints.
+As soon as it receives a RobotDesireds message, it switches the joint in the value specified in the `controlMode` attribute, either `NOT_ENABLED`,
+`POSITION_CONTROL` or `TORQUE_CONTROL`.
+When a joint enters in the `POSITION_CONTROL` state from a different state, the bridge switches its YARP control mode to `VOCAB_CM_POSITION_DIRECT`.
+Similarly when a joint enters in the `TORQUE_CONTROL` state from a different state, the bridge switches its YARP control mode to `VOCAB_CM_TORQUE`.
 
-While the bridge is in the ON state, it sends desired torques based on the control law specified in https://github.com/robotology-playground/ihmc-ors-yarp/issues/3, using data received from the RobotDesireds messages and the feedback received from the robot.  
+When a joint is in the `TORQUE_CONTROL` state, it sends desired torques based on the control law specified in https://github.com/robotology-playground/ihmc-ors-yarp/issues/3, using data received from the RobotDesireds messages and the feedback received from the robot. 
+When a joint is in the `POSITION_CONTROL` state, it sends just the `qDesired` value to the low-level position control loop.
 
-Once more then 0.1 second pass without receiving any RobotDesireds message, the bridge switches back to the OFF state, and it switches back all the joint to the YARP control mode `VOCAB_CM_POSITION`.
-
+Once more then 0.1 second pass without receiving any RobotDesireds message, the bridge switches every joint back to the `NOT_ENABLED` state, and it switches back all the joints that were not already in `NOT_ENABLED` state to the YARP control mode `VOCAB_CM_POSITION`.
 
 ## Regenerate the idl messages
 To regenerate the idl messages, you need to have the `fastrtpsgen` tool on your PATH, that is part

--- a/autogenerated/robotDesired.cxx
+++ b/autogenerated/robotDesired.cxx
@@ -33,9 +33,10 @@ using namespace eprosima::fastcdr::exception;
 
 #include <utility>
 
+
 it::iit::yarp::JointDesired::JointDesired()
 {
-    m_enabled = false;
+    m_controlMode = it::iit::yarp::NOT_ENABLED;
     m_tau = 0.0;
     m_kp = 0.0;
     m_kd = 0.0;
@@ -49,7 +50,7 @@ it::iit::yarp::JointDesired::~JointDesired()
 
 it::iit::yarp::JointDesired::JointDesired(const JointDesired &x)
 {
-    m_enabled = x.m_enabled;
+    m_controlMode = x.m_controlMode;
     m_tau = x.m_tau;
     m_kp = x.m_kp;
     m_kd = x.m_kd;
@@ -59,7 +60,7 @@ it::iit::yarp::JointDesired::JointDesired(const JointDesired &x)
 
 it::iit::yarp::JointDesired::JointDesired(JointDesired &&x)
 {
-    m_enabled = x.m_enabled;
+    m_controlMode = x.m_controlMode;
     m_tau = x.m_tau;
     m_kp = x.m_kp;
     m_kd = x.m_kd;
@@ -69,7 +70,7 @@ it::iit::yarp::JointDesired::JointDesired(JointDesired &&x)
 
 it::iit::yarp::JointDesired& it::iit::yarp::JointDesired::operator=(const JointDesired &x)
 {
-    m_enabled = x.m_enabled;
+    m_controlMode = x.m_controlMode;
     m_tau = x.m_tau;
     m_kp = x.m_kp;
     m_kd = x.m_kd;
@@ -81,7 +82,7 @@ it::iit::yarp::JointDesired& it::iit::yarp::JointDesired::operator=(const JointD
 
 it::iit::yarp::JointDesired& it::iit::yarp::JointDesired::operator=(JointDesired &&x)
 {
-    m_enabled = x.m_enabled;
+    m_controlMode = x.m_controlMode;
     m_tau = x.m_tau;
     m_kp = x.m_kp;
     m_kd = x.m_kd;
@@ -95,7 +96,7 @@ size_t it::iit::yarp::JointDesired::getMaxCdrSerializedSize(size_t current_align
 {
     size_t initial_alignment = current_alignment;
             
-    current_alignment += 1 + eprosima::fastcdr::Cdr::alignment(current_alignment, 1);
+    current_alignment += 4 + eprosima::fastcdr::Cdr::alignment(current_alignment, 4);
 
     current_alignment += 8 + eprosima::fastcdr::Cdr::alignment(current_alignment, 8);
 
@@ -115,7 +116,7 @@ size_t it::iit::yarp::JointDesired::getCdrSerializedSize(const it::iit::yarp::Jo
 {
     size_t initial_alignment = current_alignment;
             
-    current_alignment += 1 + eprosima::fastcdr::Cdr::alignment(current_alignment, 1);
+    current_alignment += 4 + eprosima::fastcdr::Cdr::alignment(current_alignment, 4);
 
     current_alignment += 8 + eprosima::fastcdr::Cdr::alignment(current_alignment, 8);
 
@@ -133,7 +134,7 @@ size_t it::iit::yarp::JointDesired::getCdrSerializedSize(const it::iit::yarp::Jo
 
 void it::iit::yarp::JointDesired::serialize(eprosima::fastcdr::Cdr &scdr) const
 {
-    scdr << m_enabled;
+    scdr << (uint32_t)m_controlMode;
     scdr << m_tau;
     scdr << m_kp;
     scdr << m_kd;
@@ -143,7 +144,7 @@ void it::iit::yarp::JointDesired::serialize(eprosima::fastcdr::Cdr &scdr) const
 
 void it::iit::yarp::JointDesired::deserialize(eprosima::fastcdr::Cdr &dcdr)
 {
-    dcdr >> m_enabled;
+    dcdr >> (uint32_t&)m_controlMode;
     dcdr >> m_tau;
     dcdr >> m_kp;
     dcdr >> m_kd;

--- a/autogenerated/robotDesired.h
+++ b/autogenerated/robotDesired.h
@@ -68,6 +68,16 @@ namespace it
         namespace yarp
         {
             /*!
+             * @brief This class represents the enumeration ORSControlMode defined by the user in the IDL file.
+             * @ingroup ROBOTDESIRED
+             */
+            enum ORSControlMode : uint32_t
+            {
+                NOT_ENABLED,
+                POSITION_CONTROL,
+                TORQUE_CONTROL
+            };
+            /*!
              * @brief This class represents the structure JointDesired defined by the user in the IDL file.
              * @ingroup ROBOTDESIRED
              */
@@ -110,30 +120,30 @@ namespace it
                 eProsima_user_DllExport JointDesired& operator=(JointDesired &&x);
                 
                 /*!
-                 * @brief This function sets a value in member enabled
-                 * @param _enabled New value for member enabled
+                 * @brief This function sets a value in member controlMode
+                 * @param _controlMode New value for member controlMode
                  */
-                inline eProsima_user_DllExport void enabled(bool _enabled)
+                inline eProsima_user_DllExport void controlMode(it::iit::yarp::ORSControlMode _controlMode)
                 {
-                    m_enabled = _enabled;
+                    m_controlMode = _controlMode;
                 }
 
                 /*!
-                 * @brief This function returns the value of member enabled
-                 * @return Value of member enabled
+                 * @brief This function returns the value of member controlMode
+                 * @return Value of member controlMode
                  */
-                inline eProsima_user_DllExport bool enabled() const
+                inline eProsima_user_DllExport it::iit::yarp::ORSControlMode controlMode() const
                 {
-                    return m_enabled;
+                    return m_controlMode;
                 }
 
                 /*!
-                 * @brief This function returns a reference to member enabled
-                 * @return Reference to member enabled
+                 * @brief This function returns a reference to member controlMode
+                 * @return Reference to member controlMode
                  */
-                inline eProsima_user_DllExport bool& enabled()
+                inline eProsima_user_DllExport it::iit::yarp::ORSControlMode& controlMode()
                 {
-                    return m_enabled;
+                    return m_controlMode;
                 }
                 /*!
                  * @brief This function sets a value in member tau
@@ -317,7 +327,7 @@ namespace it
                 eProsima_user_DllExport void serializeKey(eprosima::fastcdr::Cdr &cdr) const;
                 
             private:
-                bool m_enabled;
+                it::iit::yarp::ORSControlMode m_controlMode;
                 double m_tau;
                 double m_kp;
                 double m_kd;

--- a/autogenerated/robotDesiredPubSubTypes.cxx
+++ b/autogenerated/robotDesiredPubSubTypes.cxx
@@ -31,6 +31,7 @@ namespace it
     {
         namespace yarp
         {
+
             JointDesiredPubSubType::JointDesiredPubSubType() {
                 setName("it::iit::yarp::JointDesired");
                 m_typeSize = (uint32_t)JointDesired::getMaxCdrSerializedSize() + 4 /*encapsulation*/;

--- a/autogenerated/robotDesiredPubSubTypes.h
+++ b/autogenerated/robotDesiredPubSubTypes.h
@@ -35,6 +35,7 @@ namespace it
     {
         namespace yarp
         {
+
             /*!
              * @brief This class represents the TopicDataType of the type JointDesired defined by the user in the IDL file.
              * @ingroup ROBOTDESIRED

--- a/examples/testSocket.cpp
+++ b/examples/testSocket.cpp
@@ -69,7 +69,8 @@ int main(int argc, char* argv[])
     //Filling the vector of data to be sent to the robot
     for (size_t jnt=0; jnt < nj; jnt++)
     {
-        sender_data.jointDesireds()[jnt].tau() = 1;
+      sender_data.jointDesireds()[jnt].controlMode() = it::iit::yarp::TORQUE_CONTROL;
+      sender_data.jointDesireds()[jnt].tau() = 1;
     }
 
     while(1)

--- a/idl/robotDesired.idl
+++ b/idl/robotDesired.idl
@@ -4,9 +4,17 @@ module it
     {
         module yarp
         {
+            enum ORSControlMode
+            {
+                NOT_ENABLED,
+                POSITION_CONTROL,
+                TORQUE_CONTROL
+            };
+
+
             struct JointDesired
             {
-                boolean enabled; // Enable/disable joint
+                ORSControlMode controlMode;
                 double tau; //Desired torque
                 double kp; // Desired joint stiffness, can be zero to do only torque control
                 double kd; // Desired joint damping, can be zero to do only torque control

--- a/robots/icubGazeboSim/launch-bridgeihmcors-sim.xml
+++ b/robots/icubGazeboSim/launch-bridgeihmcors-sim.xml
@@ -53,6 +53,8 @@
          <param name="period">0.005</param>
          <param name="feedback-address">localhost</param>
          <param name="desired-address">localhost</param>
+         <!-- For testing we use a big timeout -->
+         <param name="desired-timeout">30.0</param>
 
          <action phase="startup" level="20" type="attach">
             <paramlist name="networks">

--- a/src/BridgeIHMCORS.cpp
+++ b/src/BridgeIHMCORS.cpp
@@ -14,6 +14,8 @@
 #include <climits>
 #include <cmath>
 
+
+
 namespace yarp
 {
 namespace dev
@@ -191,6 +193,7 @@ bool BridgeIHMCORS::attachWholeBodyControlBoard(const PolyDriverList& p)
             bool ok = p[devIdx]->poly->view(m_wholeBodyControlBoardInterfaces.trqs);
             ok = ok && p[devIdx]->poly->view(m_wholeBodyControlBoardInterfaces.axis);
             ok = ok && p[devIdx]->poly->view(m_wholeBodyControlBoardInterfaces.ctrlMode);
+            ok = ok && p[devIdx]->poly->view(m_wholeBodyControlBoardInterfaces.posdir);
 
             // Read number of joints to resize buffers
             int nj=0;
@@ -213,7 +216,6 @@ bool BridgeIHMCORS::attachWholeBodyControlBoard(const PolyDriverList& p)
             m_measuredControlModes.resize(nj);
             m_controlModeTorque.resize(nj, VOCAB_CM_TORQUE);
             m_controlModePosition.resize(nj, VOCAB_CM_POSITION);
-            m_desiredTorques.resize(nj);
 
             // Resize buffers
             m_robotFeedback.jointStates().resize(nj);
@@ -224,9 +226,28 @@ bool BridgeIHMCORS::attachWholeBodyControlBoard(const PolyDriverList& p)
             // Resize FastCDR feedback buffer to the size of m_robotFeedback
             m_feedbackBuffer.resize(m_robotFeedback.getCdrSerializedSize(m_robotFeedback));
 
-
             // Resize the internal buffers of the receiving thread
             m_receivingThread.resize(nj);
+
+            // Resize control mode
+            m_jointORSControlModeMutex.lock();
+            m_jointORSControlMode.resize(nj);
+            for(int jnt=0; jnt < nj; jnt++)
+            {
+                m_jointORSControlMode[jnt] = it::iit::yarp::NOT_ENABLED;
+            }
+            m_jointORSControlModeMutex.unlock();
+
+            // Reserve this buffers (they are actually filled only if necessary)
+            m_desiredControlModesJnt.reserve(nj);
+            m_desiredControlModes.reserve(nj);
+            m_desTorquesJnt.reserve(nj);
+            m_desTorques.reserve(nj);
+            m_desPosJnt.reserve(nj);
+            m_desPos.reserve(nj);
+            m_desiredControlModesJntForTimeout.reserve(nj);
+            m_desiredControlModesForTimeout.reserve(nj);
+
 
             foundDevice = ok;
             break; // see assumption
@@ -294,16 +315,45 @@ void BridgeIHMCORS::run()
         yWarning() << "BridgeIHMCORS: some sensor were not read correctly";
     }
 
-    // Handle timeout of desired messages
-    if (m_controlActive)
+    // Handle timeout of desired messages, if there are any enabled joints
+    bool areSomeJointsEnabled = false;
+    m_desiredControlModesJntForTimeout.resize(0);
+    m_desiredControlModesForTimeout.resize(0);
+    m_jointORSControlModeMutex.lock();
+    for (size_t jnt=0; jnt < m_jointTypes.size(); jnt++)
+    {
+        if (m_jointORSControlMode[jnt] != it::iit::yarp::NOT_ENABLED)
+        {
+            areSomeJointsEnabled = true;
+            m_desiredControlModesJntForTimeout.push_back(jnt);
+            m_desiredControlModesForTimeout.push_back(VOCAB_CM_POSITION);
+        }
+    }
+    m_jointORSControlModeMutex.unlock();
+
+    if (areSomeJointsEnabled)
     {
         double timeSinceLastFeedback = yarp::os::Time::now()-m_lastTimeOfReceivedDesiredMessage;
 
         if (timeSinceLastFeedback > m_desiredTimeoutInSeconds)
         {
-            m_controlActive = false;
-            m_wholeBodyControlBoardInterfaces.ctrlMode->setControlModes(m_controlModePosition.data());
-            yInfo("BridgeIHMCORS: %lf seconds passed without receiving a message from the IHMC-ORS controller, switching back the robot to position mode.", timeSinceLastFeedback);
+            m_jointORSControlModeMutex.lock();
+            for (size_t jnt=0; jnt < m_jointTypes.size(); jnt++)
+            {
+                if (m_jointORSControlMode[jnt] != it::iit::yarp::NOT_ENABLED)
+                {
+                    m_desiredControlModesJntForTimeout.push_back(jnt);
+                    m_desiredControlModesForTimeout.push_back(VOCAB_CM_POSITION);
+                    m_jointORSControlMode[jnt] = it::iit::yarp::NOT_ENABLED;
+                }
+            }
+            m_jointORSControlModeMutex.unlock();
+
+            // Warning: this is a blocking call, check if it is problematic
+            m_wholeBodyControlBoardInterfaces.ctrlMode->setControlModes(m_desiredControlModesJntForTimeout.size(),
+                                                                        m_desiredControlModesJntForTimeout.data(),
+                                                                        m_desiredControlModesForTimeout.data());
+            yInfo("BridgeIHMCORS: %lf seconds passed without receiving a message from the IHMC-ORS controller, switching controlled joints back to position mode.", timeSinceLastFeedback);
         }
     }
 }
@@ -347,7 +397,7 @@ void BridgeIHMCORS::resetInterfaces()
     m_wholeBodyControlBoardInterfaces.trqs = nullptr;
     m_wholeBodyControlBoardInterfaces.axis = nullptr;
     m_wholeBodyControlBoardInterfaces.ctrlMode = nullptr;
-    m_controlActive = false;
+    m_wholeBodyControlBoardInterfaces.posdir = nullptr;
     m_sensorsReadingsAvailable = false;
 }
 
@@ -367,39 +417,106 @@ void BridgeIHMCORS::onDesiredMessageReceived(const it::iit::yarp::RobotDesireds&
 
     // Handle control mode
 
-    // Reading the control mode is fast
+    // Reading the control mode is fast, so there is no problem in reading them
     bool needToSetControlModeToTorque = false;
     m_wholeBodyControlBoardInterfaces.ctrlMode->getControlModes(m_measuredControlModes.data());
+
+    // We check the actual control mode of each joint and save the one that need to be changed
+    // Note: a single call to setControlModes can be quite faster than several calls to setControlMode
+    m_desiredControlModesJnt.resize(0);
+    m_desiredControlModes.resize(0);
+    m_jointORSControlModeMutex.lock();
     for (size_t jnt=0; jnt < m_jointTypes.size(); jnt++)
     {
-        if (m_measuredControlModes[jnt] != VOCAB_CM_TORQUE)
+        // For each joint, we have three relevant variables
+        // * The YARP controlmode, contained in m_measuredControlModes[jnt]
+        // * The current ORS controlmode, contained in m_jointORSControlMode[jnt]
+        // * The newly requested ORS controlmode, contained in receivedRobotDesired.jointDesireds()[jnt].controlMode
+        // Depending on this variables, a new command is send to YARP or not
+        const it::iit::yarp::JointDesired& jd = receivedRobotDesired.jointDesireds()[jnt];
+
+        if ((m_measuredControlModes[jnt] != VOCAB_CM_TORQUE) && (jd.controlMode() == it::iit::yarp::TORQUE_CONTROL))
         {
-            needToSetControlModeToTorque = true;
+            m_desiredControlModesJnt.push_back(static_cast<int>(jnt));
+            m_desiredControlModes.push_back(VOCAB_CM_TORQUE);
         }
+
+        if ((m_measuredControlModes[jnt] != VOCAB_CM_POSITION_DIRECT) && (jd.controlMode() == it::iit::yarp::POSITION_CONTROL))
+        {
+            m_desiredControlModesJnt.push_back(static_cast<int>(jnt));
+            m_desiredControlModes.push_back(VOCAB_CM_POSITION_DIRECT);
+        }
+
+        // We only set a NOT_ENABLED joint in VOCAB_CM_POSITION mode only if before it was not NOT_ENABLED
+        if ((m_measuredControlModes[jnt] != VOCAB_CM_POSITION) &&
+            (jd.controlMode() == it::iit::yarp::NOT_ENABLED) &&
+            (m_jointORSControlMode[jnt] != it::iit::yarp::NOT_ENABLED))
+        {
+            m_desiredControlModesJnt.push_back(static_cast<int>(jnt));
+            m_desiredControlModes.push_back(VOCAB_CM_POSITION);
+        }
+
+        m_jointORSControlMode[jnt] = jd.controlMode();
     }
+    m_jointORSControlModeMutex.unlock();
 
     // todo(traversaro): check if this blocking call is problematic
-    if (needToSetControlModeToTorque)
+    if (m_desiredControlModesJnt.size() > 0)
     {
-        m_wholeBodyControlBoardInterfaces.ctrlMode->setControlModes(m_controlModeTorque.data());
+        m_wholeBodyControlBoardInterfaces.ctrlMode->setControlModes(m_desiredControlModesJnt.size(),
+                                                                    m_desiredControlModesJnt.data(),
+                                                                    m_desiredControlModes.data());
     }
 
     // Compute desired torques using the IHMC control law
+    // todo(traversaro) this should be moved to a separate thread so it can run at an higher rate w.r.t.
+    //                  when the desired message is received from the IHMC-ORS controller
+    m_desTorquesJnt.resize(0);
+    m_desTorques.resize(0);
+    m_desPosJnt.resize(0);
+    m_desPos.resize(0);
     m_sensorReadingsMutex.lock();
+    m_jointORSControlModeMutex.lock();
     for (size_t jnt=0; jnt < m_jointTypes.size(); jnt++)
     {
-        const it::iit::yarp::JointDesired& jd = receivedRobotDesired.jointDesireds()[jnt];
-        m_desiredTorques[jnt] = jd.tau() +
-                                jd.kp()*( jd.qDesired()  - deg2rad(m_jointPositionsFromYARPInDeg[jnt]) ) +
-                                jd.kd()*( jd.qdDesired() - deg2rad(m_jointVelocitiesFromYARPInDegPerSec[jnt]) );
+        const it::iit::yarp::JointDesired &jd = receivedRobotDesired.jointDesireds()[jnt];
+
+        if (m_jointORSControlMode[jnt] == it::iit::yarp::TORQUE_CONTROL)
+        {
+            double desTorque = jd.tau() +
+                               jd.kp() * (jd.qDesired() - deg2rad(m_jointPositionsFromYARPInDeg[jnt])) +
+                               jd.kd() * (jd.qdDesired() - deg2rad(m_jointVelocitiesFromYARPInDegPerSec[jnt]));
+            m_desTorquesJnt.push_back(static_cast<int>(jnt));
+            m_desTorques.push_back(desTorque);
+        }
+
+        if (m_jointORSControlMode[jnt] == it::iit::yarp::POSITION_CONTROL)
+        {
+            double desPosInDeg  = rad2deg(jd.qDesired());
+            m_desPosJnt.push_back(static_cast<int>(jnt));
+            m_desPos.push_back(desPosInDeg);
+        }
     }
+    m_jointORSControlModeMutex.unlock();
     m_sensorReadingsMutex.unlock();
 
-    // Send desired joint torques
-    m_wholeBodyControlBoardInterfaces.trqs->setRefTorques(m_desiredTorques.data());
+    // Send desired joint torques (if any)
+    if (m_desTorquesJnt.size() > 0)
+    {
+        m_wholeBodyControlBoardInterfaces.trqs->setRefTorques(m_desTorquesJnt.size(),
+                                                              m_desTorquesJnt.data(),
+                                                              m_desTorques.data());
+    }
 
-    // Set the state of the control to active and save the last instant in which the feedback message have been received
-    m_controlActive = true;
+    // Send desired joint positions (if any)
+    if (m_desPosJnt.size() > 0)
+    {
+        m_wholeBodyControlBoardInterfaces.posdir->setPositions(m_desPosJnt.size(),
+                                                               m_desPosJnt.data(),
+                                                               m_desPos.data());
+    }
+
+    // Save the last instant in which the feedback message have been received
     m_lastTimeOfReceivedDesiredMessage = yarp::os::Time::now();
 }
 

--- a/src/BridgeIHMCORS.h
+++ b/src/BridgeIHMCORS.h
@@ -96,6 +96,8 @@ private:
         yarp::dev::ITorqueControl        * trqs;
         yarp::dev::IAxisInfo             * axis;
         yarp::dev::IControlMode2         * ctrlMode;
+        yarp::dev::IPositionDirect       * posdir;
+
     } m_wholeBodyControlBoardInterfaces;
 
     // Helper methods to deal with C++ interfaces
@@ -110,13 +112,19 @@ private:
     std::vector<double> m_jointPositionsFromYARPInDeg;
     std::vector<double> m_jointVelocitiesFromYARPInDegPerSec;
     std::vector<double> m_jointTorquesFromYARP;
+    std::vector<int> m_desiredControlModesJntForTimeout;
+    std::vector<int> m_desiredControlModesForTimeout;
 
     // Local buffers for sending desired values, populated in onDesiredMessageReceived
-    std::vector<double> m_desiredTorques;
     std::vector<int> m_measuredControlModes;
     std::vector<int> m_controlModeTorque;
     std::vector<int> m_controlModePosition;
-
+    std::vector<int> m_desiredControlModesJnt;
+    std::vector<int> m_desiredControlModes;
+    std::vector<int> m_desTorquesJnt;
+    std::vector<double> m_desTorques;
+    std::vector<int> m_desPosJnt;
+    std::vector<double> m_desPos;
 
     // FastRTPS robot feedback message
     it::iit::yarp::RobotFeedback m_robotFeedback;
@@ -134,7 +142,8 @@ private:
 
     // Variables to handle the timeout of received messages
     double m_desiredTimeoutInSeconds;
-    std::atomic<bool> m_controlActive;
+    std::vector< it::iit::yarp::ORSControlMode > m_jointORSControlMode;
+    std::mutex m_jointORSControlModeMutex;
     std::atomic<double> m_lastTimeOfReceivedDesiredMessage;
 
 public:

--- a/src/BridgeIHMCORS.h
+++ b/src/BridgeIHMCORS.h
@@ -7,6 +7,7 @@
 
 // Standard libraries includes 
 #include <atomic>
+#include <mutex>
 #include <vector>
 
 // YARP includes
@@ -107,7 +108,7 @@ private:
     std::vector<yarp::dev::JointTypeEnum> m_jointTypes;
 
     // Local buffers for readings sensors, populated in run
-    yarp::os::Mutex m_sensorReadingsMutex;
+    std::mutex m_sensorReadingsMutex;
     std::atomic<bool> m_sensorsReadingsAvailable;
     std::vector<double> m_jointPositionsFromYARPInDeg;
     std::vector<double> m_jointVelocitiesFromYARPInDegPerSec;

--- a/src/BridgeIHMCORS.h
+++ b/src/BridgeIHMCORS.h
@@ -11,7 +11,6 @@
 #include <vector>
 
 // YARP includes
-#include <yarp/os/Mutex.h>
 #include <yarp/os/Thread.h>
 #include <yarp/os/RateThread.h>
 

--- a/tests/BridgeIHMCORSTest.cpp
+++ b/tests/BridgeIHMCORSTest.cpp
@@ -15,15 +15,37 @@
 #include <yarp/math/Rand.h>
 #include <yarp/rtf/TestCase.h>
 
+#include <iostream>
+#include <asio.hpp>
+
+// FastCDR includes for data serialization
+#include <fastcdr/FastBuffer.h>
+#include <fastcdr/Cdr.h>
+
+#include <robotDesired.h>
+#include <robotFeedback.h>
+
+
 using namespace std;
 using namespace RTF;
 using namespace yarp::os;
 using namespace yarp::sig;
 using namespace yarp::math;
+using asio::ip::udp;
+
 
 /**********************************************************************/
 class BridgeIHMCORSTest : public yarp::rtf::TestCase
 {
+    // Driver interface to the robot
+    yarp::dev::PolyDriver       m_robot;
+    yarp::dev::IEncoders       *m_iencs;
+    yarp::dev::ITorqueControl  *m_itrq;
+    yarp::dev::IAxisInfo       *m_iaxis;
+    yarp::dev::IPositionDirect *m_iposdir;
+    yarp::dev::IControlMode2   *m_ctrlmode;
+
+    int nj;
 
 public:
     /******************************************************************/
@@ -40,17 +62,197 @@ public:
     /******************************************************************/
     virtual bool setup(yarp::os::Property& property)
     {
+        yarp::os::Property deviceOptions;
+        deviceOptions.put("device","remotecontrolboardremapper");
+        Bottle axesNames;
+        Bottle & axesList = axesNames.addList();
+        axesList.addString("torso_pitch");
+        axesList.addString("torso_roll");
+        axesList.addString("torso_yaw");
+        axesList.addString("l_shoulder_pitch");
+        axesList.addString("l_shoulder_roll");
+        axesList.addString("l_shoulder_yaw");
+        axesList.addString("l_elbow");
+        axesList.addString("r_shoulder_pitch");
+        axesList.addString("r_shoulder_roll");
+        axesList.addString("r_shoulder_yaw");
+        axesList.addString("r_elbow");
+        axesList.addString("l_hip_pitch");
+        axesList.addString("l_hip_roll");
+        axesList.addString("l_hip_yaw");
+        axesList.addString("l_knee");
+        axesList.addString("l_ankle_pitch");
+        axesList.addString("l_ankle_roll");
+        axesList.addString("r_hip_pitch");
+        axesList.addString("r_hip_roll");
+        axesList.addString("r_hip_yaw");
+        axesList.addString("r_knee");
+        axesList.addString("r_ankle_pitch");
+        axesList.addString("r_ankle_roll");
+
+        deviceOptions.put("axesNames", axesNames.get(0));
+        Bottle remoteControlBoards;
+        Bottle & remoteControlBoardsList = remoteControlBoards.addList();
+        remoteControlBoardsList.addString("/icubSim/torso");
+        remoteControlBoardsList.addString("/icubSim/head");
+        remoteControlBoardsList.addString("/icubSim/left_arm");
+        remoteControlBoardsList.addString("/icubSim/right_arm");
+        remoteControlBoardsList.addString("/icubSim/left_leg");
+        remoteControlBoardsList.addString("/icubSim/right_leg");
+
+        deviceOptions.put("remoteControlBoards",remoteControlBoards.get(0));
+        deviceOptions.put("localPortPrefix","/BridgeIHMCORSTest");
+
+        Property & remoteControlBoardsOpts = deviceOptions.addGroup("REMOTE_CONTROLBOARD_OPTIONS");
+        remoteControlBoardsOpts.put("writeStrict","on");
+
+        // For the test we hardcode the considered joints
+        RTF_ASSERT_ERROR_IF_FALSE(m_robot.open(deviceOptions), Asserter::format("Unable to open robot remote_controlboard"));
+        RTF_ASSERT_ERROR_IF_FALSE(m_robot.view(m_iencs),  Asserter::format("Unable to view IEncoder interface"));
+        RTF_ASSERT_ERROR_IF_FALSE(m_robot.view(m_itrq),  Asserter::format("Unable to view ITorqueControl interface"));
+        RTF_ASSERT_ERROR_IF_FALSE(m_robot.view(m_iaxis),  Asserter::format("Unable to view IAxisInfo interface"));
+        RTF_ASSERT_ERROR_IF_FALSE(m_robot.view(m_iposdir),  Asserter::format("Unable to view IPositionDirect interface"));
+        RTF_ASSERT_ERROR_IF_FALSE(m_robot.view(m_ctrlmode),  Asserter::format("Unable to view IControlMode2 interface"));
+        RTF_ASSERT_ERROR_IF_FALSE(m_iencs->getAxes(&nj),  Asserter::format("getAxes failed"));
+
+        return true;
     }
 
     /******************************************************************/
     virtual void tearDown()
     {
+        m_robot.close();
     }
 
     /******************************************************************/
     virtual void run()
     {
-        RTF_TEST_REPORT("Computing walking trajectory");
+        RTF_TEST_REPORT("Setup UDP interface");
+
+        // Variables declaration
+        asio::io_service sender_io_service;
+        asio::io_service receiver_io_service;
+
+        udp::resolver sender_resolver(sender_io_service);
+        udp::resolver receiver_resolver(receiver_io_service);
+
+        udp::resolver::query sender_query(udp::v4(), "localhost", "9980");
+        udp::resolver::query receiver_query(udp::v4(), "localhost", "9970");
+
+        udp::endpoint sender_endpoint;
+        udp::endpoint receiver_endpoint;
+
+        // Trying to match the specified IP address and port number to an endpoint
+        sender_endpoint = *sender_resolver.resolve(sender_query);
+        receiver_endpoint = *receiver_resolver.resolve(receiver_query);
+
+        // Opening sockets
+        udp::socket sender_socket(sender_io_service);
+        udp::socket receiver_socket(receiver_io_service);
+
+        sender_socket.open(udp::v4());
+        receiver_socket.open(udp::v4());
+
+        receiver_socket.bind(receiver_endpoint);
+
+        // FastCDR buffers and serialization
+        eprosima::fastcdr::FastBuffer receiver_buffer;
+        it::iit::yarp::RobotFeedback receiver_data;
+        it::iit::yarp::RobotDesireds sender_data;
+
+        // Resize buffers
+        sender_data.jointDesireds().resize(nj);
+
+        // Receiver
+        receiver_buffer.resize(receiver_data.getMaxCdrSerializedSize());
+
+        //////////////
+        // Switch to TORQUE_CONTROL mode (VOCAB_CM_TORQUE)
+        //////////////
+
+        // Filling the vector of data to be sent to the robot
+        for (size_t jnt=0; jnt < nj; jnt++)
+        {
+            sender_data.jointDesireds()[jnt].controlMode() = it::iit::yarp::TORQUE_CONTROL;
+        }
+
+        // Send the desired message (setting the message torque)
+        eprosima::fastcdr::FastBuffer sender_buffer;
+        eprosima::fastcdr::Cdr sender_data_ser(sender_buffer);
+        sender_data.serialize(sender_data_ser);
+
+        RTF_TEST_REPORT("Send message switching to TORQUE_CONTROL");
+        sender_socket.send_to(asio::buffer(sender_buffer.getBuffer(), sender_buffer.getBufferSize()), sender_endpoint);
+
+        yarp::os::Time::delay(1.0);
+
+        // Check that the controlmodes have switched
+        std::vector<int> measuredControlModes(nj);
+
+        m_ctrlmode->getControlModes(measuredControlModes.data());
+
+        for (size_t jnt=0; jnt < nj; jnt++)
+        {
+            RTF_ASSERT_ERROR_IF_FALSE(measuredControlModes[jnt] == VOCAB_CM_TORQUE, Asserter::format("Unable to swich joint %d to VOCAB_CM_TORQUE", jnt));
+        }
+
+        //////////////
+        // Switch to POSITION_DIRECT mode (VOCAB_CM_POSITION_DIRECT)
+        //////////////
+
+        // Filling the vector of data to be sent to the robot
+        for (size_t jnt=0; jnt < nj; jnt++)
+        {
+            sender_data.jointDesireds()[jnt].controlMode() = it::iit::yarp::POSITION_CONTROL;
+        }
+
+        // Send the desired message
+        // TODO(traversaro) : cleanup all this buffers
+        eprosima::fastcdr::FastBuffer sender_buffer2;
+        eprosima::fastcdr::Cdr sender_data_ser2(sender_buffer2);
+        sender_data.serialize(sender_data_ser2);
+
+        RTF_TEST_REPORT("Send message switching to POSITION_CONTROL");
+        sender_socket.send_to(asio::buffer(sender_buffer2.getBuffer(), sender_buffer2.getBufferSize()), sender_endpoint);
+
+        yarp::os::Time::delay(1.0);
+
+        // Check that the YARP controlmodes have switched
+        m_ctrlmode->getControlModes(measuredControlModes.data());
+
+        for (size_t jnt=0; jnt < nj; jnt++)
+        {
+            RTF_ASSERT_ERROR_IF_FALSE(measuredControlModes[jnt] == VOCAB_CM_POSITION_DIRECT, Asserter::format("Unable to swich joint %d to VOCAB_CM_POSITION_DIRECT", jnt));
+        }
+
+        //////////////
+        // Switch to NOT_ENABLED mode (VOCAB_CM_POSITION)
+        //////////////
+
+        // Filling the vector of data to be sent to the robot
+        for (size_t jnt=0; jnt < nj; jnt++)
+        {
+            sender_data.jointDesireds()[jnt].controlMode() = it::iit::yarp::NOT_ENABLED;
+        }
+
+        // Send the desired message
+        RTF_TEST_REPORT("Send message switching to NOT_ENABLED");
+        // TODO(traversaro) : cleanup all this buffers
+        eprosima::fastcdr::FastBuffer sender_buffer3;
+        eprosima::fastcdr::Cdr sender_data_ser3(sender_buffer3);
+        sender_data.serialize(sender_data_ser3);
+        sender_socket.send_to(asio::buffer(sender_buffer3.getBuffer(), sender_buffer3.getBufferSize()), sender_endpoint);
+
+        yarp::os::Time::delay(1.0);
+
+        // Check that the YARP controlmodes have switched
+        m_ctrlmode->getControlModes(measuredControlModes.data());
+
+        for (size_t jnt=0; jnt < nj; jnt++)
+        {
+            RTF_ASSERT_ERROR_IF_FALSE(measuredControlModes[jnt] == VOCAB_CM_POSITION, Asserter::format("Unable to switch joint %d to VOCAB_CM_POSITION", jnt));
+        }
+
     }
 };
 

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -2,15 +2,16 @@
 # Copy Policy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
 
 # For testing we require RTF and YARP compiled with RTF support
-find_package(RTF)
-if(NOT RTF_FOUND OR NOT TARGET YARP::YARP_rtf OR NOT TARGET RTF::RTF)
+find_package(RTF 1.3.4 REQUIRED)
+if(NOT TARGET YARP::YARP_rtf)
   message(FATAL_ERROR "BUILD_TESTING is enabled, but no RTF or YARP::YARP_rtf was detected, \
                        please recompile YARP with the YARP_COMPILE_RTF_ADDONS CMake option set to ON to compile the tests in this repo.")
 endif()
 
-rtf_add_plugin(BridgeIHMCORSTest MODULE BridgeIHMCORSTest.cpp)
+rtf_add_plugin(BridgeIHMCORSTest SOURCES BridgeIHMCORSTest.cpp)
 
-target_link_libraries(BridgeIHMCORSTest PRIVATE RTF::RTF RTF::RTF_dll YARP::YARP_rtf)
+target_include_directories(BridgeIHMCORSTest PRIVATE ${Asio_INCLUDE_DIRS})
+target_link_libraries(BridgeIHMCORSTest PRIVATE RTF::RTF RTF::RTF_dll YARP::YARP_OS YARP::YARP_dev YARP::YARP_rtf fastcdr idl_messages)
 
 # Configure the testing fixture
 configure_file(${CMAKE_CURRENT_SOURCE_DIR}/fixture-sim-and-bridge.xml.in
@@ -19,6 +20,6 @@ configure_file(${CMAKE_CURRENT_SOURCE_DIR}/fixture-sim-and-bridge.xml.in
 # Add the test
 rtf_add_suite(NAME BridgeIHMCORSTestSuite
               FIXTURE yarpserver
-              FIXTURE yarpmanager
-              TEST BridgeIHMCORSTest
+              FIXTURE yarpmanager --fixture ${CMAKE_CURRENT_BINARY_DIR}/fixture-sim-and-bridge.xml
+              TEST BridgeIHMCORSTest TYPE dll
               VERBOSE)

--- a/tests/fixture-sim-and-bridge.xml.in
+++ b/tests/fixture-sim-and-bridge.xml.in
@@ -10,16 +10,15 @@
         <parameters>icub_fixed.world</parameters>
         <node>localhost</node>
     </module>
+    <!--
     <module>
         <name>gzclient</name>
-        <dependencies>
-            <port timeout="10">/icubSim/right_leg/state:o</port>
-        </dependencies>
         <node>localhost</node>
     </module>
+    -->
     <module>
         <name>yarprobotinterface</name>
-        <parameters>--config @CMAKE_SOURCE_DIR@/robots/icubGazeboSim/lauch-bridgeihmcors-sim.xml</parameters>
+        <parameters>--config @CMAKE_SOURCE_DIR@/robots/icubGazeboSim/launch-bridgeihmcors-sim.xml</parameters>
         <dependencies>
             <port timeout="10">/icubSim/right_leg/state:o</port>
         </dependencies>

